File: himura/src/main/java/be/vergauwen/simon/himura/MVPContract.java
package be.vergauwen.simon.himura;

public interface MVPContract {
    interface View {}

    interface Presenter<V extends View> {
        V getView();
        void attachView(V view);
        void detachView();
    }

    interface Component<P extends Presenter> {
        P presenter();
    }
}

File: app/src/main/kotlin/be/vergauwen/simon/mvp/MVPContract.kt
package be.vergauwen.simon.mvp

interface MVPContract {
  interface View

  interface Presenter<V : View> {
    fun getView(): V?
    fun attachView(view: View)
    fun detachView()
  }

  interface Component<V : View, P : Presenter<V>> {
    fun presenter(): P
  }
}
